lib/MusicBrainz/Server/Controller/Role/Merge.pm

```diff
@@ -194,29 +199,86 @@ role {
 
         my %entity_id = map { $_->id => $_ } @$entities;
 
-        my $new_id = $form->field('target')->value or die 'Coludnt figure out new_id';
+        my $new_id = $form->field('target')->value or die q{Couldn't figure out new_id};
         my $new = $entity_id{$new_id};
         my @old_ids = grep { $_ != $new_id } @{ $form->field('merging')->value };
 
         log_assertion { @old_ids >= 1 } 'Got at least 1 entity to merge';
 
         $c->model('MB')->with_transaction(sub {
             if ($params->edit_type) {
-                $self->_insert_edit(
-                    $c, $form,
-                    edit_type => $params->edit_type,
-                    new_entity => {
-                        id => $new->id,
-                        name => $new->name,
-                        $self->_extra_entity_data($c, $form, $new),
+                $self->edit_action($c,
+                    form        => $params->merge_form,
+                    type        => $params->edit_type,
+                    edit_args   => {
+                        new_entity => {
+                            id => $new->id,
+                            name => $new->name,
+                            $self->_extra_entity_data($c, $form, $new),
+                        },
+                        old_entities => [ map +{
+                            id => $entity_id{$_}->id,
+                            name => $entity_id{$_}->name,
+                            $self->_extra_entity_data($c, $form, $entity_id{$_}),
+                        }, @old_ids ],
+                        (map { $_->name => $_->value } $form->edit_fields),
+                        $self->_merge_parameters($c, $form, $entities),
+                    },
+                    on_creation => sub {
+                        my ($edit, $form) = @_;
+
+                        if ($c->namespace eq 'release' && $form->field('merge_rgs')->value) {
+                            my $edit_id = $edit->id;
```

Comment:
It's sometimes hard to avoid adding entity-specific code into roles, but I think in this case it should be relatively easy to add something like an empty `_merge_on_creation` method that gets called here (kinda like `_merge_parameters` or `_extra_entity_data`) that `Controller::Release` can override with all this code. That helps keep the role easier to follow as it'll mostly contain code that applies to all entities as intended.

Reply:
I did this (in a separate commit for ease of review). Tests pass but maybe it can be simplified / improved further, please have a quick check :)